compiler/passes/convert-typed-uast.cpp

```diff
@@ -0,0 +1,2041 @@
+  Symbol* sym = globalSyms[modId];
+  if (sym == nullptr) {
+    // It can happen that it wasn't in modulesToConvert but we need to
+    // convert it anyway. E.g. with PrintModuleInitOrder.
+    return setupModule(modId);
+  }
+
+  return toModuleSymbol(sym);
+}
+
+void TConverter::convertModuleInit(const Module* mod, ModuleSymbol* modSym) {
+  printf("Converting module init for %s[%i]\n",
+         mod->id().str().c_str(), modSym->id);
+
+  // create the module init function
+  modSym->initFn = new FnSymbol(astr("chpl__init_", modSym->name));
+  modSym->initFn->retType = dtVoid;
+
+  modSym->initFn->addFlag(FLAG_MODULE_INIT);
+  modSym->initFn->addFlag(FLAG_RESOLVED);
+  modSym->initFn->addFlag(FLAG_INSERT_LINE_FILE_INFO);
+
+  // add the init function to the module
+  modSym->block->insertAtHead(new DefExpr(modSym->initFn));
+
+  // TODO: handle module deinit functions
+
+  // convert module-level statements into module's init function
+
+  // prepare to traverse
+  ID id = mod->id();
+  const ResolutionResultByPostorderID& resolved = resolveModule(context, id);
+  pushBlock(modSym->initFn->body);
+  symbol = mod;
+  curFnSymbol = modSym->initFn;
+  curModuleSymbol = modSym;
+  moduleFromLibraryFile = modSym->hasFlag(FLAG_PRECOMPILED);
+  currentResolvedFunction = nullptr;
+
+  // traverse
+  ResolutionContext rcval(context);
+  ResolvedVisitor<TConverter> rv(&rcval, symbol, *this, resolved);
+  symbol->traverse(rv);
+
+  // add the 'return _void' at the end of the module init function
+  modSym->initFn->insertAtTail(new CallExpr(PRIM_RETURN, gVoid));
+
+  // tidy up after traversal
+  symbol = nullptr;
+  curFnSymbol = nullptr;
+  curModuleSymbol = nullptr;
+  popBlock();
+
+  // record any fixups that need to be handled
+  noteAllContainedFixups(modSym->initFn, 0);
+}
+
+void TConverter::convertFunction(const ResolvedFunction* r) {
+  // don't bother with functions that don't need to exist at runtime
+  if (!functionExistsAtRuntime(r)) {
+    return;
+  }
+
+  printf("Converting function %s\n", r->id().str().c_str());
+
+  // figure out, in which block should we put the DefExpr for the new FnSymbol?
+
+  // here we flatten all functions, so we just need to find the appropriate
+  // module.
+  ID parentModule = parsing::idToParentModule(context, r->id());
+  ModuleSymbol* modSym = findOrSetupModule(parentModule);
+
+  // prepare to visit
+  const ResolutionResultByPostorderID& resolved = r->resolutionById();
+  const AstNode* ast = parsing::idToAst(context, r->id());
+  if (!ast || !ast->isFunction()) {
+    CHPL_ASSERT(false && "expected Function");
+    return;
+  }
+  const Function* fn = ast->toFunction();
+  pushBlock(modSym->block);
+  symbol = fn;
+  curFnSymbol = nullptr; // to be set later
+  curModuleSymbol = nullptr; // not doing a module init here
+  currentResolvedFunction = r;
+  moduleFromLibraryFile = modSym->hasFlag(FLAG_PRECOMPILED);
+  localSyms.clear();
+
+  // traverse
+  ResolutionContext rcval(context);
```

Comment:
Here too I worry about missing resolution state from the parent function.

Reply:
Added a note to the PR's future work section.